aws/rust-runtime/aws-types/src/sdk_config.rs

```diff
@@ -153,6 +156,23 @@ impl Builder {
         self
     }
 
+    /// Set the api key to use when making requests.
+    /// # Examples
+    /// ```
+    /// use aws_types::SdkConfig;
+    /// let config = SdkConfig::builder().api_key("some-api-key").build();
+    /// ```
+    pub fn api_key(mut self, api_key: impl Into<AuthApiKey>) -> Self {
```

Comment:
We should talk more about your use-case, but I think this is probably the wrong place to add this. Adding it here will impact all AWS SDK users as this configuration will become available to them when its not technically a valid configuration for the AWS SDKs.

Essentially, right now, there are two types of smithy-rs client generation:
1. Generic client generation
2. AWS SDK client generation, which is a bundle of codegen customizations applied to (1) to add AWS SDK specific features. Basically anything in the `aws/` directory is part of this client codegen type.

HTTP auth makes sense to add directly to (1) since it's just handling the built-in http auth traits in Smithy.

Reply:
In regards to this, we are actually generating our own SDKs following very closely the pattern from the AWS SDK itself. We do that by using the parts from the Rust client codegen that are relevant for our own services.

Also, we want to use some of the default that are present in the [aws-config crate](https://crates.io/crates/aws-config). I took the approach of recreating our own config version which builds the same `SdkConfig` as defined in the [aws-types crate](https://crates.io/crates/aws-types). The reason being that we can have services with different types of authentication mechanism, one of them is actually passing the AWS provided credentials.

Reply:
What would you suggest us do? Create our own version of the `SdkConfig` containing the base options for every service we have?

Reply:
My initial idea of how this would need to work is:

- The API auth key is configurable at the smithy-rs generic client level (below the SDK level). That way, generic non-SDK clients can use HTTP auth keys like the Smithy spec calls for.
- Your client code generator adds a customization (via `ClientCodegenDecorator`--lots of examples in `sdk-codegen`) to exercise the underlying generic client configuration option.

At a high level, codegen customizations are classes on the codegen classpath that inject additional code into different hook-points. All the differences between the AWS SDK and the generic Smithy clients are done entirely via customizations (and the additional crates that get pulled in by those customizations).

Problem is: `aws-config`/`aws-types` are not customizable since they are not code generated. So any mechanism you customize in would need to be outside of those somehow. It seems hacky, but you may be able to re-use the `Credentials` type for your API key configuration, and customize converting that into the HTTP auth.

Thinking about it more though... the vast majority of the code in `aws-config` is centered around AWS credential resolution, so it's probably better to modify your code generator to drop the need for `aws-config` entirely, and rely on the generic `Client` instantiation. You could even code generate your own more convenient client construction interface if desired. Is there anything in `aws-config` that you absolutely need?

Think about the above for a bit and let me know how you want to proceed, and then I can provide more information on how to accomplish that approach.

Reply:
Your idea matches mine to the point that:
- The API auth key has been added in [this PR](https://github.com/awslabs/smithy-rs/pull/2154) (although some modification is required due to changes done here)
- We have our own codegen doing the proper customization as you described.

For now the important parts are API key, retry config, timeout config, our own concept of region, HTTP connector and endpoint resolver. Eventually we will need the AWS Credentials, but we can leave for a second wave of changes we will need to do next month.

Reply:
With that said, I will start to port the `SdkConfig` and friends to our own packages in order to eliminate this dependency and I believe I can revert the commits related to the `aws-types` crate.